src/librustc_errors/emitter.rs

```diff
@@ -749,6 +752,9 @@ fn overlaps(a1: &Annotation, a2: &Annotation) -> bool {
 fn emit_to_destination(rendered_buffer: &Vec<Vec<StyledString>>,
         lvl: &Level,
         dst: &mut Destination) -> io::Result<()> {
+    use lock;
+
+    let _buffer_lock = lock::acquire_global_lock("rustc_errors");
```

Comment:
Could you add a comment here indicating why we're acquiring a global lock here? I was also thinking that it may be best to attach this to a `Terminal` output type. If we're not emitting to a terminal (e.g. all the compile-fail tests) then there's no need to synchronize.